pydantic/mypy.py

```diff
@@ -0,0 +1,624 @@
+from mypy.util import get_unique_redefinition_name
+
+T = TypeVar('T')
+CB = Optional[Callable[[T], None]]
+
+CONFIGFILE_KEY = 'pydantic-mypy'
+METADATA_KEY = 'pydantic-mypy-metadata'
+BASEMODEL_FULLNAME = 'pydantic.main.BaseModel'
+BASESETTINGS_FULLNAME = 'pydantic.env_settings.BaseSettings'
+FIELD_FULLNAME = 'pydantic.fields.Field'
+
+
+def plugin(version: str) -> 'TypingType[Plugin]':
+    """`version` is the mypy version string"""
+    return PydanticPlugin
+
+
+class PydanticPlugin(Plugin):
+    def __init__(self, options: Options) -> None:
+        self.plugin_config = PydanticPluginConfig(options)
+        super().__init__(options)
+
+    def get_base_class_hook(self, fullname: str) -> 'CB[ClassDefContext]':
+        sym = self.lookup_fully_qualified(fullname)
+        if sym and isinstance(sym.node, TypeInfo):  # pragma: no branch
+            # No branching may occur if the mypy cache has not been cleared
+            if is_pydantic_model(sym.node):
+                return partial(pydantic_model_class_maker_callback, plugin_config=self.plugin_config)
+        return None
+
+    def get_method_hook(self, fullname: str) -> Optional[Callable[[MethodContext], Type]]:
+        if fullname.endswith('.from_orm'):
+            return from_orm_callback
+        return None
+
+
+class PydanticPluginConfig:
+    allow_required_dynamic_aliases: bool
+    allow_untyped_fields: bool
+    init_allow_extra: bool
+    init_typed: bool
+
+    def __init__(self, options: Options) -> None:
+        default_plugin_settings: Dict[str, bool] = dict(
+            allow_required_dynamic_aliases=True, allow_untyped_fields=True, init_allow_extra=True, init_typed=False
+        )
+        if options.config_file is None:  # pragma: no cover
+            return
+        plugin_config = ConfigParser()
+        plugin_config.read(options.config_file)
+        for key, default in default_plugin_settings.items():
+            setting = plugin_config.getboolean(CONFIGFILE_KEY, key, fallback=default)
+            setattr(self, key, setting)
+
+
+def is_pydantic_model(info: TypeInfo) -> bool:
+    for base in info.mro:
+        if base.fullname() == BASEMODEL_FULLNAME:
+            return True
+    return False
+
+
+def pydantic_model_class_maker_callback(ctx: ClassDefContext, plugin_config: PydanticPluginConfig) -> None:
+    transformer = PydanticModelTransformer(ctx, plugin_config)
+    transformer.transform()
+
+
+def from_orm_callback(ctx: MethodContext) -> Type:
+    """Raise an error if orm_mode is not enabled"""
+    model_type: Instance
+    if isinstance(ctx.type, CallableType) and isinstance(ctx.type.ret_type, Instance):
+        model_type = ctx.type.ret_type  # called on the class
+    elif isinstance(ctx.type, Instance):
+        model_type = ctx.type  # called on an instance (unusual, but still valid)
+    else:  # pragma: no cover
+        # I'm not sure if there is a way to get to this branch, but if so, it is very rare
```

Comment:
One of the mypy devs told me it can be something else, but didn't tell me what else it can be! I'll add a little convenience function that says please create a github issue to let us know you ran into this fork.